doc/sonic-build-system/SONiC-Reproduceable-Build.md

```diff
@@ -0,0 +1,460 @@
+
+To publish only the packages used by SONiC, the used packages and versions should be collected by SONiC build. The input file should be like this as below:
+
+_Automat==0.6.0_
+
+_Babel==2.3.4_
+
+_Babel==2.6.0_
+
+_Flask==1.1.2_
+
+_Jinja2==2.10_
+
+_Jinja2==2.11.2_
+
+_Jinja2==2.7.3_
+
+For one package, it may contain several versions to publish. The input file schema is the same as the Pypi version control file.
+
+The [python-pypi-mirror](https://pypi.org/project/python-pypi-mirror/) is used to publish the Pypi mirror to [Azure Storage Static Website](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blob-static-website), likes Debian package mirror, it supports CDN and traffic manager.
+
+# Docker Base Images
+
+In the original design, the base images of the SONiC Dockerfile are based on image tag, for instance, in the Dockerfile sonic-slave-buster/Dockerfile, the base image is debian:buster, we can find &quot;From debian:buster&quot; in the file. When a new debian buster image is pushed in the docker registry, the debian:buster will refer to the new one, it is not reproduceable. To make it support reproduceable, we can change to refer to the digest hash value of the image. The docker base images that are based on the image tag will be changed to base on the image digest to support reproduceable build. The Dockerfile as below is based on the image tag:
+
+_From <image>:<tag>_
+
+Example:
+
+_From debian:buster_
+
+Change to:
+
+_From <image>:<digest>_
+
+Example:
+
+_From debian@sha256:439a6bae1ef351ba9308fc9a5e69ff7754c14516f6be8ca26975fb564cb7fb76_
+
+The docker file will be changed to use the right digest during the build based on a version control file in file/build/versions/versions-docker-base, the content of the file is as below:
+
+debian:buster==_sha256:439a6bae1ef351ba9308fc9a5e69ff7754c14516f6be8ca26975fb564cb7fb76_
+
+debian:stretch==_sha256:335ecf9e8d9b2206c2e9e7f8b09547faa9f868e694f7c5be14c38be15ea8a7cf_
+
+debian:jessie==_sha256:8fc7649643ca1acd3940706613ea7b170762cfce6e7955a6afb387aa40e9f9ea_
+
+# Git Repository
+
+The SONiC build will download source code from remote repository during the build. For instance, when building sflow, it will download source code by command: git clone [https://github.com/sflow/sflowtool](https://github.com/sflow/sflowtool). It will download the latest source code, not reproduceable.
+
+A version control file is introduced to control which commit of the git repository will be download when building SONiC image. The schema of the version control file is same as wget/curl. The content of the control file likes as below:
+
+[https://github.com/sflow/sflowtool==1e42bc69fab8a8bc58b04a0ff9ad2c366eef6121](https://github.com/sflow/sflowtool==1e42bc69fab8a8bc58b04a0ff9ad2c366eef6121)

```

Comment:
This link is so misleading because it is non-existing URL. I propose a new format as:
```
https://github.com/sflow/sflowtool,master,1e42bc69fab8a8bc58b04a0ff9ad2c366eef6121
```
It includes fields:
1. original cloned repo
2. original checked out branch name
3. the commit id used for reproducible build, it is fixed when we build it from repo and branch at the first time

I notice github and salsa.debian.org are using different url schema, so no need to follow any. In theory this whole line may still be a misleading url, but less misleading.

Reply:
we can change the format to "<url>,<commit>", the branch name is no use now, we can add it when we use it.
https://github.com/sflow/sflowtool,1e42bc69fab8a8bc58b04a0ff9ad2c366eef6121

Reply:
Currently, we are using the same format. In most time, may not need to change it by human.